public/main.js

```diff
@@ -48,24 +48,101 @@ form.onsubmit = function(event) {
-    createRequest.open("POST", "/api/todo");
-    createRequest.setRequestHeader("Content-type", "application/json");
-    createRequest.send(JSON.stringify({
-        title: title,
+//the flag parameter is set when the response body should be use for the callback
+function responseHandlerFactory(callback, err, status, flag) {
+    return function(res) {
+        if (res.status === status) {
+            if (flag) {
+                res.json().then(callback);
+            }
+            else {
+                callback();
+            }
+        }
+        else {
+            res.text().then(function(data) {
+                error.textContent = err + res.status + " - " + data;
+            });
+        }
+    };
+}
+
+function createTodo(str, callback) {
+    var payload = {
+        title: str,
         complete: false
-    }));
-    createRequest.onload = onLoadFactory(createRequest, "Failed to create item. Server returned ",
-                           201, callback);
+    };
+
+    fetch("/api/todo", {
+        method: "POST",
+        headers: {
+            "Content-type": "application/json"
+        },
+        body: JSON.stringify(payload)
+    })
+    .then(
+        responseHandlerFactory(callback, "Failed to create item. Server returned ", 201, false)
+    );
 }
 
 function getTodoList(callback) {
-    var createRequest = new XMLHttpRequest();
-    createRequest.open("GET", "/api/todo");
-    createRequest.onload = onLoadFactory(createRequest, "Failed to get list. Server returned ",
-                           200, callback, JSON.parse, "responseText");
-    createRequest.send();
+
+    fetch("/api/todo", {
+        method: "GET",
+        headers: {
+            "Content-type": "application/json"
+        }
+    })
+    .then(
+        responseHandlerFactory(callback, "Failed to get list. Server returned ", 200, true)
+    );
+}
+
+function completeEntry(todo) {
+    var payload = {
+        id: todo.id,
+        complete: true
+    };
+    fetch("/api/todo/", {
+        method: "PUT",
+        headers: {
+            "Content-type": "application/json"
+        },
+        body: JSON.stringify(payload)
+    })
+    .then(
+        responseHandlerFactory(reloadTodoList, "Failed to update. Server returned ", 200, false)
+    );
+}
+
+function deleteList(todoList) {
+    var executed = 0;
+    var listener = function() {
+        if (executed < 2) {
+            reloadTodoList();
+        }
+        else {
+            executed--;
+        }
+    };
+    for (var i = 0; i < todoList.length; i++) {
+        executed++;
+        deleteEntry(todoList[i], listener);
+    }
+}
+
+function deleteEntry(todo, callback) {
+    callback = callback || reloadTodoList;
+
+    fetch("/api/todo/" + todo.id, {
+        method: "DELETE",
+        headers: {
+            "Content-type": "application/json"
+        }
+    })
+    .then(
+        responseHandlerFactory(callback, "Failed to delete. Server returned ", 200, false)
```

Comment:
There's a `catch` method on the fetch api so you can split these here something like
```
fetch()
    .then(handleData)
    .catch(handleError)
```